lib/recursive-methods.rb

```diff
@@ -1,49 +1,141 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) since n operations will be needed to compute factorial(n)
+# Space complexity: O(n) since each recursive call will end up with a stack frame on the call stack, so n in total for factorial(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    if n < 0
+        raise ArgumentError
+    elsif n == 0 || n == 1
+        return 1
+    else
+        return n*factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) the slicing method is O(n) itself, within another O(n) method
+# Space complexity: O(n^2) the slicing method is O(n) itself, within another O(n) method
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s == nil || s.length == 0
+        return ""
+    else
+        return reverse(s[1..-1]) + s[0]
+    end
+end
+
+# Time complexity: O(n) the reverse_inplace method without the helper, is O(1),
+# the if statement is O(n), and the recursion is O(n), they are not nested, so we are still at O(n)
+# Space complexity: O(n) because we have as many stack calls as we have steps
+def reverse_inplace_helper(s, start_position, end_position)
+    if start_position >= end_position
+        return
+    end
+
+    temp = s[start_position]
+    s[start_position] = s[end_position]
+    s[end_position] = temp
+
+    reverse_inplace_helper(s, start_position+1, end_position-1 )
 end
 
-# Time complexity: ?
-# Space complexity: ?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    if s == nil || s.length == 0
+        return ""
+    else
+        reverse_inplace_helper(s, 0, s.length - 1)
+        return s
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because we will make n iterations in total for bunny(n)
+# Space complexity: O(n) because we will make 1 stack call per step, i.e. n in total for bunny(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 2 if n == 1
+    return 0 if n == 0
+    return 2 + bunny(n-1)
+end
+
+# Time complexity: O(n) because there is the actual method has complexity of O(1) without its helper
+# and the helper has no nesting, just the recursion which will take place n times
+# Space complexity: O(n) same reasoning because for every step, we have one stack call
+def nested_helper(s, start_position, end_position, left, right)
```

Comment:
This works on the tests, but you **really** should not be using instance variables here.  

Also you are counting if the number of "(" equals the number of ")".  What about  ")("?